subsys/bluetooth/host/gatt.c

```diff
@@ -2542,20 +2594,140 @@ int bt_gatt_notify_cb(struct bt_conn *conn,
 int bt_gatt_notify_multiple(struct bt_conn *conn, uint16_t num_params,
 			    struct bt_gatt_notify_params *params)
 {
-	int i, ret;
+	int i;
+	int err;
+	size_t total_len = 0;
+	struct net_buf *buf;
+	struct notify_data data;
 
 	__ASSERT(params, "invalid parameters\n");
-	__ASSERT(num_params, "invalid parameters\n");
 	__ASSERT(params->attr, "invalid parameters\n");
 
+	/* Send any outstanding notifications */
+	gatt_notify_flush(conn);
+
+	CHECKIF(num_params < 2) {
```

Comment:
Minor: This function is rather long now (around 140 lines). Consider moving all the parameter validation to a separate function.

Reply:
agreed, but I'm having trouble seeing the limit between parameter and data validation. like does [the feature support](https://github.com/jori-nordic/zephyr/blob/e1c0e2665228d7142d0cc61f9db762847e49a199/subsys/bluetooth/host/gatt.c#L2646) count as a parameter or does [the total PDU length](https://github.com/jori-nordic/zephyr/blob/e1c0e2665228d7142d0cc61f9db762847e49a199/subsys/bluetooth/host/gatt.c#L2646) also count ?
I would also think that the [callback functions](https://github.com/jori-nordic/zephyr/blob/e1c0e2665228d7142d0cc61f9db762847e49a199/subsys/bluetooth/host/gatt.c#L2665) are considered as function parameters.

If I had to decide somewhere to split, it would be right before allocating the netbuf. What do you think about that ?

Reply:
> If I had to decide somewhere to split, it would be right before allocating the netbuf. What do you think about that ?

Sounds like a fine place to do it. 

w.r.t. splitting the function, we can also do something like

```c

static bool gatt_notify_multiple_valid_args(...)
{
    ...
}

static bool gatt_notify_multiple_valid_params(...)
{
    ...
}

static bool gatt_notify_multiple_create_pdu(...)
{
    ...
}

int bt_gatt_notify_multiple(struct bt_conn *conn, uint16_t num_params,
			    struct bt_gatt_notify_params *params)
{
    if (gatt_notify_multiple_valid_args()) {
        return -EINVAL;
    }

    if (gatt_notify_multiple_valid_params()) {
        return -EINVAL;
    }

    buf = gatt_notify_multiple_create_pdu();
    if (buf == NULL) {
        return -EINVAL;
    }
        
    return gatt_notify_mult_send(conn, buf);
}
```

Or have the functions return `int` (errno) values instead. 

The point is that we can easily add multiple functions for parameter validation etc., and if they are only used once (like they will here) they are going to be inlined by the compiler and thus we shouldn't see any performance/memory difference.

Reply:
Yeah I did this, but kept the PDU building in the main function, as two different error codes could be returned. I think it's clearer now.